crates/sui-types/src/transaction.rs

```diff
@@ -239,6 +241,19 @@ impl CallArg {
                     mutable,
                 }]
             }
+            // Receiving objects are not part of the input objects.
+            CallArg::Object(ObjectArg::Receiving(_)) => vec![],
+        }
+    }
+
+    fn receiving_objects(&self) -> Vec<ObjectRef> {
```

Comment:
What's the reason to not include the Receiving in the input objects with a new InputObjectKind, and handle this new kind differently when needed? Asking because I believe there are much more similarities between input objects and receiving objects, than differences.

Reply:
My first crack at implementing this is exactly as you describe (received objects are returned as input objects with a new `InputObjectKind`). That direction is definitely workable I believe, however what I encountered when doing that was there were a lot of places where `InputObjectKinds` flow where `Receiving` objects make no sense (e.g., in checking or loading object arguments to a transaction). So this led to a bunch of `unreachable!`/`invariant_violation` code. @tnowacki had some opinions on this I believe as well (or maybe I'm misremembering). 

Moving to separating them out seemed like a cleaner option since it led to significantly less unreachable/invariant_violation code paths and (IMO) seemed to fit more. I actually view `Receiving` objects as more akin to dynamic fields than input objects other than the addition of dependencies, but I'm more than happy to also prototype treating them as input objects some more and we can take a look at both options and decide if you'd like :)

Reply:
I think the main thing with not treating `Receiving` objects as input objects is there are also a number of places where they need to be treated differently from other input objects, especially around the fact that we (1) don't lock them like we do with other objects, and (2) we don't check their existence prior to execution -- in this way they're more akin to dynamic fields with a specified version than an input object. Additionally at the execution layer, `Receiving` objects and other input objects differ significantly in terms of how they should be treated and checked.

Because of this the worry is that if we treat `Receiving` objects as input objects, they could accidentally leak to someplace they should and we might, e.g., accidentally grab a lock on them (or try to load them etc.) when we shouldn't. Treating them separately makes this much less likely IMO.